File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.java
import java.util.*;

class Solution {
    public String solution(int n, int m, int x, int y, int r, int c, int k) {
        String answer = "";        
        String[] dirStr = new String[]{"d","l","r","u"};
        int[][] dirs = {{1,0},{0, -1},{0,1},{-1,0}};

        int realk = Math.abs(r-x) + Math.abs(c-y);
        if(realk == k){
            answer = "d".repeat(Math.max(r-x, 0)) + "l".repeat(Math.max(-(c-y), 0)) + "r".repeat(Math.max(c-y,0)) + "u".repeat(Math.max(-(r-x),0));
            return answer;
        } else if ((k-realk) % 2 == 1 || k < realk){
            return "impossible";
        } else {
            int cx = x, cy = y;
            StringBuilder sb = new StringBuilder();
            while (k > 0) {
                for (int i = 0; i < 4; i++) {
                    int nx = cx + dirs[i][0], ny = cy + dirs[i][1];
                    if (nx <= 0 || nx > n || ny <= 0 || ny > m) {
                        continue;
                    }
                    int distance = Math.abs(r - nx) + Math.abs(c - ny);
                    if (distance <= k - 1) {
                        k--;
                        cx = nx;
                        cy = ny;
                        sb.append(dirStr[i]);
                        break;
                    }
                }
            }
            return sb.toString();
        }
    }
    
}

File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.kt
import java.util.*
import kotlin.math.abs
class Solution {
fun solution(n: Int, m: Int, x: Int, y: Int, r: Int, c: Int, k: Int): String {
    var answer: String = ""
    val distance = abs(x-r) + abs(y-c)
    if(k < distance || (k - distance) % 2 > 0){
        return "impossible"
    }

//    val p = ((k-distance)/2) + 1
//    var visited = Array(n+1) { IntArray(m+1) { p } }
    val dirs = arrayOf(
        Pair("d", Pair(1, 0)), // down
        Pair("l", Pair(0, -1)), // left
        Pair("r", Pair(0, 1)), // right
        Pair("u", Pair(-1, 0)) // up
    )
    val queue = LinkedList<Node>() // 현재까지 위치 누적, x,y

    queue.add(Node(x,y, listOf()))

    while(queue.isNotEmpty()){
        var now = queue.poll()

        for(d in dirs){
            val nx = now.x + d.second.first
            val ny = now.y + d.second.second

            if(nx <= 0 || nx > n || ny <= 0 || ny > m) continue
            val left = abs(nx - r) + abs(ny - c)
            if(left > k - (now.d.size + 1) || (k-(now.d.size + 1)-left) % 2 > 0) continue
//            if(visited[nx][ny] <= 0) continue
            if((now.d.size + 1) == k){
                if(nx == r && ny == c){
                    return now.d.joinToString("") + d.first
                } else continue
            }
//            visited[nx][ny] -= 1
//            println("nd = $nd, nx = $nx, ny = $ny")
            val nd = now.d.plus(d.first)
            queue.add(Node(nx, ny, nd))
            break
        }

    }
    return answer
}


}

data class Node(val x: Int, val y : Int, val d : List<String>){

}